composer/callbacks/mlperf.py

```diff
@@ -0,0 +1,285 @@
+            raise FileExistsError(f'{filename} already exists.')
+
+        self._file_handler = logging.FileHandler(filename)
+        self._file_handler.setLevel(logging.INFO)
+        self.mllogger.logger.addHandler(self._file_handler)
+
+        # TODO: implement cache clearing
+        self.mllogger.start(key=mllog.constants.CACHE_CLEAR)
+        self.mllogger.start(key=mllog.constants.INIT_START)
+
+        if rank_zero():
+            self._log_dict({
+                constants.SUBMISSION_BENCHMARK: benchmark,
+                constants.SUBMISSION_DIVISION: division,
+                constants.SUBMISSION_ORG: submitter,
+                constants.SUBMISSION_PLATFORM: system_name,
+                constants.SUBMISSION_STATUS: status,
+            })
+
+    def _create_submission_folders(self, root_folder: str, system_name: str, benchmark: str):
+        if not os.path.isdir(root_folder):
+            raise FileNotFoundError(f"{root_folder} not found.")
+
+        results_folder = os.path.join(root_folder, 'results')
+        log_folder = os.path.join(root_folder, 'results', system_name)
+        benchmark_folder = os.path.join(log_folder, benchmark)
+        systems_folder = os.path.join(root_folder, 'systems')
+
+        os.makedirs(results_folder, exist_ok=True)
+        os.makedirs(log_folder, exist_ok=True)
+        os.makedirs(benchmark_folder, exist_ok=True)
+        os.makedirs(systems_folder, exist_ok=True)
+
+    def _log_dict(self, data: Dict):
+        for key, value in data.items():
+            self.mllogger.event(key=key, value=value)
+
+    def fit_start(self, state: State, logger: Logger) -> None:
+        if rank_zero():
+            if state.train_dataloader.batch_size is None:
+                raise ValueError("Batch size is required to be set for dataloader.")
+
+            self._log_dict({
+                constants.SEED: state.seed,
+                constants.GLOBAL_BATCH_SIZE: state.train_dataloader.batch_size * dist.get_world_size(),
+                constants.GRADIENT_ACCUMULATION_STEPS: state.grad_accum,
+                constants.TRAIN_SAMPLES: len(state.train_dataloader.dataset),
+                constants.EVAL_SAMPLES: len(state.evaluators[0].dataloader.dataloader.dataset)
+            })
+
+        self.mllogger.event(key=constants.INIT_STOP)
+
+        dist.barrier()
+        if rank_zero():
+            self.mllogger.event(key=constants.RUN_START)
+
+    def epoch_start(self, state: State, logger: Logger) -> None:
+        if rank_zero():
+            self.mllogger.event(key=constants.EPOCH_START, metadata={'epoch_num': state.timer.epoch.value})
+            self.mllogger.event(key=constants.BLOCK_START,
+                                metadata={
+                                    'first_epoch_num': state.timer.epoch.value,
+                                    'epoch_count': 1
+                                })
+
+    def epoch_end(self, state: State, logger: Logger) -> None:
+        if rank_zero():
```

Comment:
is it OK if every rank logs, and then we only submit the rank zero folder?

Reply:
The MLperf logger logs from multiple ranks into the same file, specifcally the cache_clear, init_start, and init_end events -- there should be as many of these for each rank. For an example results file, see [NVIDIA submission](https://github.com/mlcommons/training_results_v1.1/blob/main/NVIDIA/results/dgxa100_n128_ngc21.09_pytorch/bert/result_0.txt).

Reply:
The NV/Google submission code logs directly to the same file, leading me to believe that this is unneeded code complexity, probably because only <8 entries are concurrently written. Let's keep as-is and can fix later if we run into an issue.